libdrm doesn't build against musl (regression)

I'm in the process of upgrading `static-haskell-nix` to latest nixpkgs `master` (commit 07ce48cc3a1c6551a754ad487d2d9a90361530e5), and get the following build failure for `libdrm`:

```
[58/87] Compiling C object 'tests/nouveau/e47a46e@@threaded@exe/threaded.c.o'.
FAILED: tests/nouveau/e47a46e@@threaded@exe/threaded.c.o 
/nix/store/ad3j1a9myvk162458qpj8qzw14kdmm2w-gcc-wrapper-7.4.0/bin/cc -Itests/nouveau/e47a46e@@threaded@exe -Itests/nouveau -I../tests/nouveau -I. -I../ -I../include/drm -Itests/nouveau>
../tests/nouveau/threaded.c:39:5: error: conflicting types for ‘ioctl’
 int ioctl(int fd, unsigned long request, ...)
     ^~~~~
In file included from ../tests/nouveau/threaded.c:23:0:
/nix/store/dd1il3gf9q3386pk68kik56mn52pvs7j-musl-1.1.22-dev/include/sys/ioctl.h:119:5: note: previous declaration of ‘ioctl’ was here
 int ioctl (int, int, ...);
     ^~~~~
ninja: build stopped: subcommand failed.
```

The problem didn't occur before commit 4c77d471e40ae1b1581cdbc33fb8fae00e6fd0b7 `libdrm: build with meson` (from #57773) because the tests weren't built by default with autotools. See
https://github.com/openembedded/openembedded-core/commit/8dcebad54ded4fa9b3455b674be37ad0b10190d8#diff-4295180ffa50d118533fe15edfd3c729 which says

> As Meson always builds the test suite (instead of building before running it),
add a patch to fix compilation with musl.

The compile problem seems to be that glibc decided to not implement the POSIX type of `ioctl()`, but to change the second parameter from `int` to `unsigned long`).

Most projects `#ifdef` that problem:

* https://bugs.freedesktop.org/show_bug.cgi?id=85319
* https://patchwork.linuxtv.org/patch/28041/

Other distros carry patches for `libdrm` because of that:

* OpenEmbedded: https://github.com/openembedded/openembedded-core/blob/master/meta/recipes-graphics/drm/libdrm/musl-ioctl.patch
* Alpine Linux: https://git.alpinelinux.org/aports/commit/main/libdrm/ioctl.patch?id=7ba3523fe098c050e2729a4cd9a1fd50e2167aca
* Gentoo: https://gitweb.gentoo.org/proj/musl.git/diff/x11-libs/libdrm/files/fix-ioctl.patch?id2=5600c0735d024b65c378a86f36a9d128d2111fb6

Now that we use meson, we should most likely also carry such a patch.

CC @colemickens @dtzWill

--- a/pkgs/development/libraries/libdrm/default.nix
+++ b/pkgs/development/libraries/libdrm/default.nix
@@ -1,5 +1,5 @@
 { stdenv, lib, fetchurl, pkgconfig, meson, ninja, libpthreadstubs, libpciaccess
-, withValgrind ? valgrind-light.meta.available, valgrind-light
+, withValgrind ? valgrind-light.meta.available, valgrind-light, fetchpatch
 }:
 
 stdenv.mkDerivation rec {
@@ -17,7 +17,15 @@ stdenv.mkDerivation rec {
   buildInputs = [ libpthreadstubs libpciaccess ]
     ++ lib.optional withValgrind valgrind-light;
 
-  patches = [ ./cross-build-nm-path.patch ];
+  patches = [ ./cross-build-nm-path.patch ] ++
+    lib.optionals stdenv.hostPlatform.isMusl [
+      # Fix tests not building on musl because they use the glibc-specific
+      # (non-POSIX) `ioctl()` type signature. See #66441.
+      (fetchpatch {
+        url = "https://raw.githubusercontent.com/openembedded/openembedded-core/30a2af80f5f8c8ddf0f619e4f50451b02baa22dd/meta/recipes-graphics/drm/libdrm/musl-ioctl.patch";
+        sha256 = "0rdmh4k5kb80hhk1sdhlil30yf0s8d8w0fnq0hzyvw3ir1mki3by";
+      })
+    ];
 
   postPatch = ''
     for a in */*-symbol-check ; do

